services/ymax-planner/src/support.ts

```diff
@@ -102,3 +102,125 @@ export const createEVMContext = async ({
     usdcAddresses: usdcAddresses[clusterName],
   };
 };
+
+type BinarySearch = {
+  (
+    start: number,
+    end: number,
+    isAcceptable: (idx: number) => Promise<boolean> | boolean,
+  ): Promise<number>;
+  (
+    start: bigint,
+    end: bigint,
+    isAcceptable: (idx: bigint) => Promise<boolean> | boolean,
+  ): Promise<bigint>;
+};
+
+export const binarySearch = (async <Index extends number | bigint>(
+  start: Index,
+  end: Index,
+  isAcceptable: (idx: Index) => Promise<boolean> | boolean,
+): Promise<Index> => {
+  const nonzero: Index = (start || ~start) as Index;
+  const unit: Index = (nonzero / nonzero) as Index; // i.e., 1 or 1n
```

Comment:
Having slept on this, I now think simpler is better.
```suggestion
  const unit = (typeof start === 'bigint' ? 1n : 1) as Index;
```